### PR TITLE
Bluetooth: host: Improve error reporting on gatt subscribe

### PR DESCRIPTION
**Pull request by Yagoor**
* Return EALREADY when the subscription already exists but it is a different entry. This indicates to the application to not wait on subscribe callback. This is needed because the callback is called from the ccc_write which does not happen when another subscription exists

### CODE REVIEW

**Comment by Yagoor** [comment]
> Since this is a change in public API behavior, shouldn't it be documented in the release notes or migration guide as well?

If the change is accepted, I can document it.

**Comment by jhedberg** [comment]
> > Since this is a change in public API behavior, shouldn't it be documented in the release notes or migration guide as well?
> 
> If the change is accepted, I can document it.

It seems reasonable to me. The migration guid would likely be the right place for it.

**Review by jhedberg — DISMISSED** [review]

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -5541,7 +5541,11 @@ int bt_gatt_subscribe(struct bt_conn *conn,
 	 */
 	sys_slist_prepend(&sub->list, &params->node);
 
-	return 0;
+	/*
+	 * Report already if subscription exist so that application does not wait
+	 * for the subscribe callback
+	 */
+	return has_subscription ? -EALREADY : 0;
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:5548]: I'm not 100% certain this is the right approach. 

In this case we may return an error, while still adding the params to the list. That means this is a case where we return an error while still changing the state, which I'm not a fan off. 

Did you consider the approach of calling the callback instead without the GATT write when it exists? (Although I'm not a fan of that approach either tbh).

Comment (Yagoor) [reply subsys/bluetooth/host/gatt.c:5548]: I see your dilemma, it sounds wrong to change the state and still return an error.

I don't have a problem with that approach either. 
The disadvantage in that case is that we will be indicating to the application that a ccc write happened. I am not sure if this can have any implication to some application.

**Comment by alwa-nordic** [comment]
Alternative: Invoke the callback, emulating a successful response.

File: doc/releases/migration-guide-4.0.rst [hunk 2]
```diff
@@ -404,6 +404,9 @@ Bluetooth Audio
 Bluetooth Host
 ==============
 
+* ``bt_gatt_subscribe`` returns ``-EALREADY`` when another subscription exists. This indicates to
```
Comment (Thalley) [inline doc/releases/migration-guide-4.0.rst:407]:
```suggestion
* :c:func:`bt_gatt_subscribe` returns ``-EALREADY`` when another subscription exists. This indicates to
```

**Comment by jhedberg** [comment]
Discussed in the Bluetooth WG. Consensus seems (for now) to be toward the "safe approach" i.e. just do another write over the air, as proposed by @Thalley here: https://github.com/zephyrproject-rtos/zephyr/pull/82107#discussion_r1861953147

File: subsys/bluetooth/host/gatt.c (hunk 1 above)
Comment (Yagoor) [reply subsys/bluetooth/host/gatt.c:5548]: Yeah, my assumption is that it is also a optimization.
I'll give it a try, but my assumption is that it will work without a problem.

**Review by Thalley — COMMENTED** [review]
Changes look good to me, although I'm unaware if we have any tests that tests this?

**Comment by Yagoor** [comment]
> Changes look good to me, although I'm unaware if we have any tests that tests this?

The subscription is indirectly tested in many bsims (specially the ones that involve audio)
So, the current behavior should not have changed if those tests are satisfied.

**Review by jhedberg — DISMISSED** [review]

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
We must consider the case when the application subscribes to indications and then to notifications. I don't know what the original intention behind the API is for this case, but from just reading the code, I see the behavior has changed.

Before this change, it results in indications. After this change, it results in notifications.

**Comment by Yagoor** [comment]
> We must consider the case when the application subscribes to indications and then to notifications. I don't know what the original intention behind the API is for this case, but from just reading the code, I see the behavior has changed.
> 
> Before this change, it results in indications. After this change, it results in notifications.

Good catch! If I propagate the value, it should be good right?

Application subscribes to indications and then to notifications.
Before this change, it results in indications.
After it would result in indications and notifications.

**Comment by Thalley** [comment]
> The server may reject a request to set both the Notification and Indication bits for the same characteristic.

In nearly all GATT requests the server may reject the request for any reason, so this is a weird thing to write here. A GATT server may even reject a read request to the device name if it chooses. 

In any case, if a GATT client user in Zephyr chooses to set both bits, then that's up to that user. 

I think I'm missing the detail that you are discussing. Isn't the subscription value something that the client writes to the server, and then it is up to the server to decide what to do here? Are we, as the client, assuming some behavior?

**Comment by alwa-nordic** [comment]
I posted the quote from the spec because I think @Yagoor is asking whether he should update the code to set both bits when we have two subscriptions from the application, where one subscription is asking for notifications and the other is asking for indications. I think this is a subtly difficult question to answer. But we can put that discussion aside for now.

To maintain the previous behavior w.r.t. the value of the CCC, the value must be "indications" when the application subscribes to indications and notifications separately, and it must be "indications and notifications" if the application is requesting that in any single subscription.

In other words, we must bring back the loop removed by this PR, but instead of computing `has_subscription`, we compute `ccc_value = MAX(all subscriptions, including the new one)`.

**Comment by Thalley** [comment]
> In other words, we must bring back the loop removed by this PR, but instead of computing `has_subscription`, we compute `ccc_value = MAX(all subscriptions, including the new one)`.

Ah, I understand it now. 

So if application A has subscribed to indications and then application B later subscribes to notifications, then we (in current version) overwrite application A's value. 

Not sure what your meant by MAX here, but shouldn't it be the union of all subscriptions? 

So if A subscribes to indications and B later subscribes to notifications, then we send `BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE`?

File: subsys/bluetooth/host/gatt.c [hunk 3]
```diff
@@ -5467,26 +5467,20 @@ int bt_gatt_subscribe(struct bt_conn *conn,
 		}
 
 		/* Check if another subscription exists */
-		if (tmp->value_handle == params->value_handle &&
-		    tmp->value >= params->value) {
-			has_subscription = true;
+		if (tmp->value_handle == params->value_handle) {
+			params->value |= tmp->value;
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:5471]: We should actually probably add a check for `params->value` to be between 0x01 and 0x03. 

E.g. 
App A calls `bt_gatt_subscribe` with value = 0x04 (not a valid value). We do not reject this, and add App A's param to the `sub->list`

While that request is outstanding, App B calls `bt_gatt_subscribe` with 0x01 (notify), but with this change, the value sent to the server become 0x05 which is also invalid. 

App A would receive a write error from the server because it sends a wrong value, but so will App B, even though it provided a valid value. 

Checking that `params->value` is a valid value before sending the request and before doing `sys_slist_prepend`

**Review by Thalley — CHANGES_REQUESTED** [review]

**Comment by alwa-nordic** [comment]
> Not sure what your meant by MAX here

I mean the maximum given the ordering `(notify & indicate) > indicate > notify > 0`. It just happens that the CCC values for those are 3, 2, 1, 0, respectively, so the numerical maximum will give the same result. This is a little hidden and counts as smart-programming, I would not be against something more explicit.

> shouldn't it be the union of all subscriptions?
> 
> So if A subscribes to indications and B later subscribes to notifications, then we send `BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE`?

The old code does `>=`.

https://github.com/zephyrproject-rtos/zephyr/blob/2f583a8369ef5b67c4438c40d17721d31db952a0/subsys/bluetooth/host/gatt.c#L5471
 
If we want to retain the old behavior, we must take the maximum. In your example, we send `BT_GATT_CCC_INDICATE`.

If someone wants to change the behavior, please say so, and please state your reasoning. I'm just assuming we want to keep old behavior until we intentionally decide to change it.

**Comment by Thalley** [comment]
> If someone wants to change the behavior, please say so, and please state your reasoning. I'm just assuming we want to keep old behavior until we intentionally decide to change it.

Semantically I want to change it. It does not make sense to treat indications as a superset of notifications. If I subscribe to notifications and then my callback gives me the indication value, then I assume there's a bug. 

Practically I don't think this will ever be an issue to be honest. Characteristics that have both indication and notifications are corner cases, and Zephyr applications where we want to subscribe to both, but individually, is equally much a corner case, so this ends up being a corner case squared :D

**Comment by Thalley** [comment]
> Should I rollback to the simple change in which we just send the write even if another subscription exists?

No I think the current solution is better. 

```
if existing ccc == 0 (none) and params->value is 0 => Send 0
if existing ccc == 0 (none) and params->value is 1 => Send 1
if existing ccc == 0 (none) and params->value is 2 => Send 2
if existing ccc == 0 (none) and params->value is 3 => Send 3
if existing ccc == 1 (none) and params->value is 0 => Send 1
if existing ccc == 1 (none) and params->value is 1 => Send 1
if existing ccc == 1 (none) and params->value is 2 => Send 3
if existing ccc == 1 (none) and params->value is 3 => Send 3
if existing ccc == 2 (none) and params->value is 0 => Send 2
if existing ccc == 2 (none) and params->value is 1 => Send 3
if existing ccc == 2 (none) and params->value is 2 => Send 2
if existing ccc == 2 (none) and params->value is 3 => Send 3
if existing ccc == 3 (none) and params->value is 0 => Send 3
if existing ccc == 3 (none) and params->value is 1 => Send 3
if existing ccc == 3 (none) and params->value is 2 => Send 3
if existing ccc == 3 (none) and params->value is 3 => Send 3
```

It should however be documented that if a different user/application has already subscribed to the same handle, then the callback will contain the "sum" (or max or whatever) of all the users/applications. From a GATT client application standpoint, it shouldn't matter (unless perhaps for a few GATT services) whether a notification or indication was received.

File: subsys/bluetooth/host/gatt.c [hunk 4]
```diff
@@ -5446,6 +5446,11 @@ int bt_gatt_subscribe(struct bt_conn *conn,
 		return -ENOTCONN;
 	}
 
+	if (params->value < BT_GATT_CCC_NOTIFY
+		|| params->value > BT_GATT_CCC_INDICATE) {
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:5450]: `params->value` may also be `BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE`.

```suggestion
	if (!IN_RANGE(params->value, BT_GATT_CCC_NOTIFY, BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE) {
```

We could also consider defining a mask for this purpose

File: doc/releases/migration-guide-4.0.rst [hunk 5]
```diff
@@ -404,6 +404,10 @@ Bluetooth Audio
 Bluetooth Host
 ==============
 
+* :c:func:`bt_gatt_subscribe` behavior has changed to if a different user/application has already
+  subscribed to the same handle, then the callback will contain the values aggregated of all the
+  users/applications.
```
Comment (Thalley) [inline doc/releases/migration-guide-4.0.rst:409]: This change also needs to added to the documentation of `bt_gatt_subscribe` if not already

**Review by Thalley — CHANGES_REQUESTED** [review]

File: doc/releases/migration-guide-4.0.rst (hunk 5 above)
Comment (Thalley) [reply doc/releases/migration-guide-4.0.rst:409]: @Yagoor please be careful about resolving review comments yourself: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

> Unless they applied the reviewer’s recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it.

File: include/zephyr/bluetooth/gatt.h [hunk 6]
```diff
@@ -2076,6 +2076,9 @@ struct bt_gatt_subscribe_params {
  *  notified value. One may then decide whether to unsubscribe directly from
  *  this callback. Notification callback with NULL data will not be called if
  *  subscription was removed by this method.
+ *  If a different user/application has already subscribed to the same handle,
+ *  then the callback will contain the values aggregated of all the
+ *  users/applications.
```
Comment (Thalley) [inline include/zephyr/bluetooth/gatt.h:2081]: Would it make sense to have an example of this? 

There are 2 callbacks involved here. 
One is the `bt_gatt_subscribe_params.notify` which contains the data as well as the params when we receive a notification or indication. The callback itself does not tell the application if it received an indication or notification (which is probably should have). 

The other is the `bt_gatt_subscribe_params.subscribe` which is basically the write response. This just gives back the params. 

So none of the callbacks actually contain information about whether it was a notification or indication. 

But the documentation you've added here is wrong since the callback does not contain information about the type. What can happen is that App A subscribes to notifications and App B subscribes to indications, and looking at `att_notify`, `att_indicate` and `bt_gatt_notification`, these are just handled in the same way. So even if App A only subscribes to notifications, it may receive indications, or vice versa. 

This is what the documentation should say, unless we want to take it a step further and change the behavior so that App A only gets notifications and App B only gets indications (so that the callback behavior matches the params->value). The change should be fairly trivial.

@jhedberg @alwa-nordic what are you thoughts about this? 

And sorry that this is becoming a bigger thing that you probably intended @Yagoor :D

**Review by Thalley — CHANGES_REQUESTED** [review]

File: include/zephyr/bluetooth/gatt.h (hunk 6 above)
Comment (Yagoor) [reply include/zephyr/bluetooth/gatt.h:2081]: I'll hold on any change until this is settled :)

Comment (jhedberg) [reply include/zephyr/bluetooth/gatt.h:2081]: The proposal from @Thalley sounds reasonable to me

Comment (alwa-nordic) [reply include/zephyr/bluetooth/gatt.h:2081]: > change the behavior so that App A only gets notifications and App B only gets indications (so that the callback behavior matches the params->value). The change should be fairly trivial.

It feel to me like we are just making random changes to our API without any use cases in mind. If it was true that notifications and indications were truly independent, this would be an obvious change. But the spec is being confusing again. I just want to understand what API the profiles will need.

**Comment by Thalley** [comment]
@Yagoor will you fix the failing compliance check?

**Comment by Yagoor** [comment]
> @Yagoor will you fix the failing compliance check?

Sorry, this fell through the cracks.
Should I go ahead with your proposal?

**Comment by Thalley** [comment]
> > @Yagoor will you fix the failing compliance check?
> 
> Sorry, this fell through the cracks. Should I go ahead with your proposal?

Yeah I think that makes sense, to avoid additional changes in behavior

File: include/zephyr/bluetooth/gatt.h (hunk 6 above)
Comment (Yagoor) [reply include/zephyr/bluetooth/gatt.h:2093]: I tried to improve the migration guide and I also moved it to 4.1, since 4.0 is already out.

Comment (Thalley) [reply include/zephyr/bluetooth/gatt.h:2093]: I think it makes sense to have both places. The migration guide are nice to read to see what has changed, but it doesn't count as documentation for the API

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Comment by Thalley** [comment]
@alwa-nordic please re-review or dismiss you change requests

File: subsys/bluetooth/host/gatt.c [hunk 7]
```diff
@@ -5494,26 +5499,20 @@ int bt_gatt_subscribe(struct bt_conn *conn,
 		}
 
 		/* Check if another subscription exists */
-		if (tmp->value_handle == params->value_handle &&
-		    tmp->value >= params->value) {
-			has_subscription = true;
+		if (tmp->value_handle == params->value_handle) {
+			params->value |= tmp->value;
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/gatt.c:5503]: What is the intended function of this code?

It just does not seem sensible to me to modify the subscription that is being added based on previous subscriptions. Why should we do it? The code comment "Check if another subscription exists" does not explain what happens here.

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
The description of the PR and the commit message and patch seem to have diverged. Can you clarify what the stated goal of this PR is?

AFAIU this is an overhaul of how subscriptions work in Zephyr. We to make it possible to know when a subscription is acknowledged (as stated in the PR description), we want to still allow multiple subscriptions to the same attribute, and we want to allow a mixed subscription with both notifications and indications.

The previous behavior and the proposed behavior both have problems. I think we need to ensure the system in always in a consistent state. Consider what happens in the following sequence:

 - bt_gatt_subscribe(notification)
 - bt_gatt_subscribe(indication)
 - bt_gatt_unsubscribe(indication)

At least with the implementation in this PR, this will leave the system in an inconsistent state. Remote will have `0` in the CCC, while local will have a notification subscription.

I seem to remember we agreed at some point (in the Bluetooth WG meeting?) to not allow mixed subscriptions? I think that would decrease the complexity of the needed change a lot.
